fdbserver/BulkLoadUtil.actor.cpp

```diff
@@ -0,0 +1,195 @@
+
+std::string generateRandomBulkLoadDataFileName() {
+	return deterministicRandom()->randomUniqueID().toString() + "-data.sst";
+}
+
+std::string generateRandomBulkLoadBytesSampleFileName() {
+	return deterministicRandom()->randomUniqueID().toString() + "-bytesample.sst";
+}
+
+ACTOR Future<Optional<BulkLoadState>> getBulkLoadState(Database cx, UID dataMoveId, UID logId) {
+	// TODO(Zhe): check if bulk load task is outdated
+	loop {
+		state Transaction tr(cx);
+		try {
+			Optional<Value> val = wait(tr.get(dataMoveKeyFor(dataMoveId)));
+			if (!val.present()) {
+				TraceEvent(SevWarn, "SSBulkLoadDataMoveIdNotExist", logId).detail("DataMoveID", dataMoveId);
+				return Optional<BulkLoadState>(); // TODO(Zhe): is this case not expected?
+			}
+			DataMoveMetaData dataMoveMetaData = decodeDataMoveValue(val.get());
+			return dataMoveMetaData.bulkLoadState;
+		} catch (Error& e) {
+			wait(tr.onError(e));
+		}
+	}
+}
+
+ACTOR Future<BulkLoadState> updateBulkLoadTaskPhase(Transaction* tr, KeyRange range, UID taskId, BulkLoadPhase phase) {
+	state BulkLoadState bulkLoadState;
+	RangeResult result = wait(krmGetRanges(tr, bulkLoadPrefix, range));
+	if (result.size() > 2) {
+		throw bulkload_task_outdated();
+	} else if (result[0].value.empty()) {
+		throw bulkload_task_outdated();
+	}
+	bulkLoadState = decodeBulkLoadState(result[0].value);
+	ASSERT(bulkLoadState.getTaskId().isValid());
+	ASSERT(bulkLoadState.getRange() == KeyRangeRef(result[0].key, result[1].key));
+	if (taskId != bulkLoadState.getTaskId() || bulkLoadState.phase == BulkLoadPhase::Complete) {
+		// The taskId persisted on disk can be only updated by users
+		// We ignore any complete task
+		throw bulkload_task_outdated();
+	}
+	bulkLoadState.phase = phase;
+	return bulkLoadState;
+}
+
+void bulkLoadFileCopy(std::string fromFile, std::string toFile, size_t fileBytesMax) {
+	std::string content = readFileBytes(fromFile, fileBytesMax);
+	writeFile(toFile, content);
+	// Do file checksum for toFile
+	return;
+}
+
+ACTOR Future<SSBulkLoadFileSet> bulkLoadTransportCP_impl(std::string dir,
+                                                         BulkLoadState bulkLoadState,
+                                                         size_t fileBytesMax,
+                                                         UID logId) {
+	ASSERT(bulkLoadState.getTransportMethod() == BulkLoadTransportMethod::CP);
+	loop {
+		state std::string toFile;
+		state std::string fromFile;
+		state SSBulkLoadFileSet fileSet;
+		try {
+			fileSet.folder = abspath(joinPath(dir, bulkLoadState.getFolder()));
+
+			// Clear existing folder
+			platform::eraseDirectoryRecursive(fileSet.folder);
+			if (!platform::createDirectory(fileSet.folder)) {
+				throw retry();
+			}
+
+			// Move bulk load files to loading folder
+			for (const auto& filePath : bulkLoadState.getDataFiles()) {
+				fromFile = abspath(filePath);
+				toFile = abspath(joinPath(fileSet.folder, generateRandomBulkLoadDataFileName()));
```

Comment:
Should the generated file name have a relation to the input file name? Will make it easier to trace files across the system?

Reply:
Here I intentionally decouple the name binding so that the following code logic is decoupled with the input file name.